File: generated_code/python/solution_4220_2.py
from itertools import pairwise

class Solution:
    def minOperations(self, s: str) -> int:
        # Time Complexity: O(N) where N is the length of s, as we only perform linear scans.
        # Space Complexity: O(1) auxiliary space.
        
        # Case 1: The string is already fully sorted.
        if all(a <= b for a, b in pairwise(s)):
            return 0
            
        n = len(s)
        # Case 2: If the string has length 2 and is not sorted, we cannot sort it
        # because the only valid operation requires selecting a proper substring (length < 2),
        # which means individual characters cannot be reordered.
        if n == 2:
            return -1
            
        # Identify the smallest and largest characters present in the entire string.
        mn = min(s)
        mx = max(s)
        
        # Case 3: If the first element is already the minimum character or the 
        # last element is already the maximum character, we can sort the remaining 
        # n-1 characters in a single operation.
        if s[0] == mn or s[-1] == mx:
            return 1
            
        # Case 5: If the absolute minimum element is stuck at the last position AND
        # the absolute maximum element is stuck at the first position, it takes 3 operations.
        # Example: s = "d...a" where 'a' is min and 'd' is max.
        # 1. Sort suffix excluding s[0] -> "da..."
        # 2. Sort prefix excluding s[-1] -> "ad..." (moves 'a' to front)
        # 3. Sort suffix excluding s[0] -> "a...d" (moves 'd' to end)
        # We check for this by tracking the first occurrence of min and last occurrence of max.
        min_idx = s.find(mn)
        max_idx = s.rfind(mx)
        if min_idx == n - 1 and max_idx == 0:
            return 3
            
        # Case 4: For all other configurations, we can always sort the string in 2 operations.
        return 2

File: generated_code/cpp/solution_4220_2.cpp
#include <string>
#include <algorithm>
#include <vector>

using namespace std;

class Solution {
public:
    int minOperations(string s) {
        // A string of length 0 or 1 is already sorted.
        if (s.length() <= 1) return 0;
        
        // Check if the string is already sorted.
        bool is_sorted = true;
        for (size_t i = 1; i < s.length(); ++i) {
            if (s[i] < s[i - 1]) {
                is_sorted = false;
                break;
            }
        }
        if (is_sorted) return 0;
        
        // Find the target sorted version of the string.
        string sorted_s = s;
        sort(sorted_s.begin(), sorted_s.end());
        
        // If the first character of s matches the smallest character,
        // we can sort the suffix s[1...n-1] to sort the whole string.
        // Since s is not the entire string, this takes 1 operation.
        if (s.front() == sorted_s.front()) return 1;
        
        // If the last character of s matches the largest character,
        // we can sort the prefix s[0...n-2] to sort the whole string.
        if (s.back() == sorted_s.back()) return 1;
        
        // If the first character of s is the largest and the last character is the smallest,
        // we cannot do it in 1 or 2 operations directly if we need to swap their roles,
        // but we can always do it in 3 operations:
        // 1. Sort s[0...n-2]: moves the smallest element away from the end.
        // 2. Sort s[1...n-1]: moves the largest element away from the front.
        // 3. Sort s[0...n-2]: finalizes the sort.
        // However, we must check if 2 operations are possible. 
        // 2 operations are possible if the first character is NOT the maximum character
        // AND the last character is NOT the minimum character. Since we already checked
        // front == min and back == max, 2 operations are possible if front != max OR back != min.
        if (s.front() == sorted_s.back() && s.back() == sorted_s.front()) {
            return 3;
        }
        
        // Otherwise, we can achieve it in 2 operations:
        // e.g., sort s[0...n-2] then sort s[1...n-1].
        return 2;
    }
};